pallets/cash/src/internal/lock.rs

```diff
@@ -1,167 +1,86 @@
-
-    let chain_id = holder.chain_id();
-    let chain_cash_principal_new = sub_principal_amounts(
-        ChainCashPrincipals::get(chain_id),
-        principal,
-        Reason::InsufficientChainCash,
-    )?;
-    let holder_cash_principal_new = holder_cash_principal.add_amount(principal)?;
-    let total_cash_principal_new = sub_principal_amounts(
-        TotalCashPrincipal::get(),
-        holder_repay_principal,
-        Reason::RepayTooMuch,
-    )?;
+    CashPipeline::new()
+        .lock_cash::<T>(recipient, principal)?
+        .commit::<T>();
 
     let index: CashIndex = GlobalCashIndex::get(); // Grab cash index just for event
-    ChainCashPrincipals::insert(chain_id, chain_cash_principal_new);
-    CashPrincipals::insert(holder, holder_cash_principal_new);
-    TotalCashPrincipal::put(total_cash_principal_new);
-
-    <Module<T>>::deposit_event(Event::LockedCash(sender, holder, principal, index));
+    <Module<T>>::deposit_event(Event::LockedCash(sender, recipient, principal, index));
 
     Ok(())
 }
 
+// TODO: Test
 pub fn undo_lock_internal<T: Config>(
     asset: AssetInfo,
     sender: ChainAccount,
-    holder: ChainAccount,
-    amount: AssetQuantity,
+    recipient: ChainAccount,
+    quantity: AssetQuantity,
 ) -> Result<(), Reason> {
-    // XXX undo effects as best as possible...
-    // let holder_asset = AssetBalances::get(asset.asset, holder);
-    // let (holder_repay_amount, holder_supply_amount) = repay_and_supply_amount(holder_asset, amount);
-
-    // let holder_asset_new = add_amount_to_balance(holder_asset, amount)?;
-    // let total_supply_new =
-    //     add_amount_to_raw(TotalSupplyAssets::get(asset.asset), holder_supply_amount)?;
-    // let total_borrow_new = sub_amount_from_raw(
-    //     TotalBorrowAssets::get(asset.asset),
-    //     holder_repay_amount,
-    //     Reason::RepayTooMuch,
-    // )?;
-
-    // let (cash_principal_post, last_index_post) = core::effect_of_asset_interest_internal(
-    //     asset,
-    //     holder,
-    //     holder_asset,
-    //     holder_asset_new,
-    //     CashPrincipals::get(holder),
-    // )?;
-
-    // LastIndices::insert(asset.asset, holder, last_index_post);
-    // CashPrincipals::insert(holder, cash_principal_post);
-    // TotalSupplyAssets::insert(asset.asset, total_supply_new);
-    // TotalBorrowAssets::insert(asset.asset, total_borrow_new);
-
-    // core::set_asset_balance_internal::<T>(asset.asset, holder, holder_asset_new);
-
-    // <Module<T>>::deposit_event(Event::UnLocked(asset.asset, sender, holder, amount.value));
+    // Note: we don't check liquidity here since who knows
+    CashPipeline::new()
+        .extract_asset::<T>(recipient, asset.asset, quantity)?
+        .commit::<T>();
+
+    <Module<T>>::deposit_event(Event::ReorgRevertLocked(
+        asset.asset,
+        sender,
+        recipient,
+        quantity.value,
+    ));
 
     Ok(())
 }
 
+// TODO: Test
```

Comment:
Not certain about this, but this implementation should be pretty close to what we want.